File: agclib/MicrophoneController.cpp
#include <stdexcept>

#include <mmdeviceapi.h>
#include <endpointvolume.h>

#include "COM.h"
#include "MicrophoneController.h"

namespace agc {

MicrophoneController::MicrophoneController()
{
	ComPtr<IMMDeviceEnumerator> deviceEnumerator;
	comcheck(CoCreateInstance(
		__uuidof(MMDeviceEnumerator),
		NULL,
		CLSCTX_INPROC_SERVER,
		__uuidof(IMMDeviceEnumerator),
		reinterpret_cast<LPVOID*>(deviceEnumerator.put())));

	ComPtr<IMMDevice> endpointDevice;
	comcheck(deviceEnumerator->GetDefaultAudioEndpoint(eCapture, eConsole, endpointDevice.put()));

	comcheck(endpointDevice->Activate(
		__uuidof(IAudioEndpointVolume),
		CLSCTX_ALL,
		NULL,
		reinterpret_cast<void**>(m_audioEndpointVolume.put())));
}

void MicrophoneController::setVolume(float volume)
{
	if (volume < 0.0f || volume > 1.0f) {
		throw std::invalid_argument("volume must be in the range [0.0, 1.0]");
	}
	comcheck(m_audioEndpointVolume->SetMasterVolumeLevelScalar(volume, NULL));
}

float MicrophoneController::getVolume() const
{
	float volume;
	comcheck(m_audioEndpointVolume->GetMasterVolumeLevelScalar(&volume));
	return volume;
}

}

File: agclib/src/microphone_controller.rs
use windows::Win32::{
    Media::Audio::{
        eCapture, eConsole,
        Endpoints::IAudioEndpointVolume,
        IMMDeviceEnumerator, MMDeviceEnumerator,
    },
    System::Com::{CoCreateInstance, CLSCTX_ALL, CLSCTX_INPROC_SERVER},
};

use crate::{AgcError, Result};

/// Controls the scalar volume of the default audio recording endpoint.
///
/// The underlying COM interface is reference-counted by `windows-rs`; `Drop`
/// releases it automatically — do not call `Release` manually.
pub struct MicrophoneController {
    endpoint_volume: IAudioEndpointVolume,
}

impl MicrophoneController {
    pub fn new() -> Result<Self> {
        // SAFETY: COM must be initialised on this thread (via ComGuard) before
        // calling CoCreateInstance. The caller is responsible for that invariant.
        let endpoint_volume = unsafe {
            let enumerator: IMMDeviceEnumerator =
                CoCreateInstance(&MMDeviceEnumerator, None, CLSCTX_INPROC_SERVER)?;
            let device = enumerator.GetDefaultAudioEndpoint(eCapture, eConsole)?;
            device.Activate(CLSCTX_ALL, None)?
        };
        Ok(Self { endpoint_volume })
    }

    /// Sets the master scalar volume.
    ///
    /// # Errors
    ///
    /// Returns [`AgcError::InvalidVolume`] if `volume` is outside `[0.0, 1.0]`.
    pub fn set_volume(&self, volume: f32) -> Result<()> {
        if !(0.0_f32..=1.0_f32).contains(&volume) {
            return Err(AgcError::InvalidVolume(volume));
        }
        // SAFETY: endpoint_volume is a valid COM interface acquired in new().
        unsafe {
            self.endpoint_volume
                .SetMasterVolumeLevelScalar(volume, std::ptr::null())?;
        }
        Ok(())
    }

    /// Returns the current master scalar volume in `[0.0, 1.0]`.
    pub fn get_volume(&self) -> Result<f32> {
        // SAFETY: endpoint_volume is a valid COM interface acquired in new().
        unsafe { Ok(self.endpoint_volume.GetMasterVolumeLevelScalar()?) }
    }
}
